### ISSUE TITLE
Add remove() method to ConanFile as entry point for uninstalling system_requirements

### ISSUE THREAD

**Issue by Aalmann** [conan-io/conan#2774]
Hi,

here is a new feature request.

## Motivation ##
Conan provides a `system_requirements()` method to install system-wide software (e.g. from package repository). We are using this (on Windows) to call installer packages (e.g. Mathworks Matlab, Visual Studio, etc.) which can't be packaged and zipped as "normal" packages could (like CMake).

## Problem ##
Currently there is no (good) way to call the uninstaller. So the user has to uninstall the software manually (e.g. by running apt-get on Linux or running the uninstaller link on Windows). Calling `conan remove <packagename>` only deletes the directory in local cache but does not perform any further actions.

## Feature request ##
A new method `remove()` should be added to conan file to be able to run an uninstaller. This method has to be called when running `conan remove`. Additionally there should be an optional argument (e.g. `-sm, --skip-method`) at `conan remove` to skip the method call. The method itself should have the same options (like os_info and SystemPackageTools) as `system_requirements()`.

Thanks.
Best Aalmann

To help us debug your issue please explain:

- [x] I've read the [CONTRIBUTING guide](https://raw.githubusercontent.com/conan-io/conan/develop/.github/CONTRIBUTING.md).
- [x] I've specified the Conan version, operating system version and any tool that can be relevant.
- [x] I've explained the steps to reproduce the error or the motivation/use case of the question/suggestion.

**Comment by lasote** [comment]
I understand the motivation, but for me the `system_requirements()` is something to avoid if possible. I don't think we should start managing the uninstalling too, it would require having a registry of what was really installed in the system to not uninstall more than the needed, and for sure will be corner cases. I see it too much hard and risky for the value it brings.

**Comment by memsharded** [comment]
Yes, it would require some kind of "reference" counting. What happens if 2 recipes installed something, and you remove one of them? You cannot remove that system thing, because the other recipe would stop working, right?

**Comment by Aalmann** [comment]
So what I want to say is, you can't avoid that users break their systems by doing crazy things which they are probably not aware of. This could always happen (e.g. they uninstall something manually which was installed by a recipe). And it will always happen because it is python and one can do everything with it.

The `system_requirements()` method is a really good entry point which can be used to reproducible install a whole system installed tool chain. We are using it to completely setup a clean dev client with all the tools required to work properly by just calling `conan install <toolchain_name>`. But currently there is no official way to cleanup the system installations in an easy way. And that's why I request this feature.

**Comment by Aalmann** [comment]
Hi @memsharded and @lasote 

Isn't it possible to just add this 'optional' method and the additional flags to `conan remove` with a big hint in documentation that it could cause problems when uninstalling things which could also be installed by another recipe?! 
I agree, having a registry would be nice and probably useful for a later iteration. But that "global registry" is already missing for `system_requirements`. I have seen that there is a `system_reqs` directory created in the local cache which contains a `system_reqs.txt` file. But this file just seems to be filled when using a `SystemPackageTool` and the file is not "globally stored". 

So please discuss with the team. The feature would help a lot.

Best Aalmann

**Comment by Aalmann** [comment]
@lasote and @memsharded 
Would be nice to get this feature (just the callback of `remove()` within the recipe) in 1.5.

Thanks.
Best Aalmann

**Comment by Aalmann** [comment]
@lasote
Is there a way to get this feature soon?
We are currently creating a PoC using Chocolatey for system wide installations. This will also be used to install special compiler toolchains using a profile which contains `build_requirement` recipe which will use `system_requirements` which will trigger `choco install` to install the special compiler. 
So it would be very nice to have also a `remove_system_requirements` method in recipe which can be used for cleanup. 

Best Aalmann

**Comment by lasote** [comment]
We still don't see a `remove_system_requirements` supported by Conan. Because N packages (a, b, c, d) can install the same system requirement and probably you don't want to uninstall it when you only uninstall one of them (a). Even with different caches in CI could be a mess.

We can consider to add a callback to the [plugins system](https://docs.conan.io/en/latest/reference/plugins.html) for deteted recipes, we could pass the loaded conanfile and let you call a custom method that you could call it `remove_system_requirements` ;). WDYT?

**Comment by Aalmann** [comment]
@lasote 
Because of your example with `a,b,c and d` I suggested the `--sm` flag to skip the method execution if not required / desired (and only if implemented). 
But you will always have this problem and you still have it with the current conan solution. Let's say, that `a,b,c,d` require system installation (e.g. qt5-qmake or something else, there are a lot of other packages available in different versions) but `d` uses a backported higher version. You can not make sure, that the same package in different versions will work with all packages (it should work, but you never can't make sure). Depending on installation order `a,b,c,d` or `a,b,d,c` or `d,c,b,a` you may get different system installation results and probably different errors and the user can't figure out what's going on. So in this case the best solution is to define additional package `e` which is the only package containing the `system_requirements` and all other packages depending on `e`. Then you have exactly one single point of system installation.

Having this combination in CI could always be a big problem (e.g. there is/was a problem with Visual Studios PDB server executable when running builds in parallel for different projects/solutions). That's why we only use one single build executor on each slave/agent and try to run build/make in parallel within the same solution. And that's only when using a system wide installed tool. 

However, getting it as plugin system callback would be fine. Even if I don't see the "real" difference and advantage comparing to the/my optional method approach in `ConanFile`.

Best Aalmann

**Comment by memsharded** [comment]
The ``remove`` system_requirements never gained traction, because of the dangerous destructive approach, and because the default ``mode`` of Conan changed to be only checking, but not installing system dependencies. In that default case, removal from recipes would also be quite unexpected, and useless in the majority of cases.

Conan 2.0 has some relevant features regarding this, that can probably facilitate some custom tasks:

- It is possible to collect the system_requirements with ``conf`` like ``tools.system.package_manager:mode=report``, and that can be easily collected from json output
- There are custom commands that could easily build a graph, collect the system_requirements, then do whatever action they want, like uninstalling.

So at the moment, adding a method to the recipes for the removal of system packages is not in the roadmap, closing this ticket as not planned. Please report or ask any question on the new features if necessary.
